tensorboard/data/dispatching_provider_test.py

```diff
@@ -0,0 +1,337 @@
+# Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Unit tests for `tensorboard.data.dispatching_provider`."""
+
+import base64
+
+from tensorboard import errors
+from tensorboard import context
+from tensorboard import test as tb_test
+from tensorboard.data import dispatching_provider
+from tensorboard.data import provider
+
+
+class FiniteProvider(provider.DataProvider):
```

Comment:
I’m curious to hear what reviewers think of this test pattern. My
thoughts:

  - I like that we’re testing the observable behavior of the real code
    with valid data providers, rather than relying on mocking or spying
    or anything brittle like that.
  - I considered using the existing multiplexer provider as sub-provider
    for these tests, but it felt inappropriate because the multiplexer
    provider doesn’t care about experiment IDs at all (not to mention
    that it touches disk).
  - Because the data provider interface is fairly large (many methods,
    many parameters), this test provider is fairly long, even though we
    skip tensors entirely.
  - In general, I like to introduce a kind of “configurable in-memory
    implementation” for things like this, so that in the test code you
    can say “add an experiment with this ID and this kind of data”. We
    don’t yet have one of those for the `DataProvider` interface itself.
    Would now be a good time to add one?

Overall, I think that these tests are reasonably robust, but I’m open to
ideas on how to improve them, along any axis.

Reply:
FWIW, optional especially since it's merged already but I think the "assert equality to the bare provider response" approach would also work for the blobs, right?  Since the only part of the response being altered is the blob keys, but those are invisible to the assertion since we feed them back into the same provider.

Reply:
Right again; thanks! Sent #3858 (but entirely non-urgent).